### PR TITLE
[TM DownloadFile Pause and Resume] Part 3: Implement resumeDownloadFile

### PR DESCRIPTION
**Pull request by zoewangg**
## Modifications
Implement `S3TransferManager#resumeDownloadFile`. 

Below is the workflow for resume
```
1. Send HeadObjectRequest to retrieve the last modified time of the S3 object
2. Compare s3ObjectLastModified and fileLastModified to see if the file or the s3 object has been modified since last pause.
3. If either of them has been modified, the SDK will download the object from the beginning and replace the existing content in the file.
4. If neither of them has been modified, the SDK will start to download from the next byte
```

## Testing
Added unit tests and functional tests

### CODE REVIEW

File: services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/ResumableFileDownload.java [hunk 1]
```diff
@@ -66,31 +64,25 @@ public boolean equals(Object o) {
...
     }
 
-    @Override
-    public String toString() {
-        return ToString.builder("ResumableFileDownload")
-                       .add("downloadFileRequest", downloadFileRequest)
-                       .add("bytesTransferred", bytesTransferred)
-                       .add("lastModified", lastModified)
-                       .add("transferSizeInBytes", transferSizeInBytes)
-                       .build();
```
Comment (zoewangg) [inline services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/ResumableFileDownload.java:91]: This method will be implemented in a separate PR as part of serialization/deserialization implementation

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/FileTransformerConfiguration.java [hunk 2]
```diff
@@ -79,11 +79,11 @@ public static FileTransformerConfiguration defaultCreateNew() {
...
      * <p>
      * Create a new file if it doesn't exist, otherwise replace the existing file.
-     * In the event of an error, the SDK will attempt to delete the file (whatever has been written to it so far).
+     * In the event of an error, the SDK will NOT attempt to delete the file, leaving it as-is
      */
     public static FileTransformerConfiguration defaultCreateOrReplaceExisting() {
         return builder().fileWriteOption(FileWriteOption.CREATE_OR_REPLACE_EXISTING)
-                        .failureBehavior(FailureBehavior.DELETE)
+                        .failureBehavior(FailureBehavior.LEAVE)
                         .build();
```
[the remark this answers is not in the document]
Comment (dagnir) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/FileTransformerConfiguration.java:87]: No I think this behavior is fine

File: services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/ObjectTransfer.java [hunk 3]
```diff
@@ -31,7 +32,7 @@
 @SdkPreviewApi
 public interface ObjectTransfer extends Transfer {
     /**
-     * The stateful {@link TransferProgress} associated with this transfer.
+     * The {@link CompletableFuture} holding the stateful {@link TransferProgress} associated with this transfer.
      */
-    TransferProgress progress();
+    CompletableFuture<TransferProgress> progress();
```
[the remark this answers is not in the document]
Comment (zoewangg) [reply services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/ObjectTransfer.java:37]: I tried it, but later realized that`TransferProgress` is a stateful object (`TrasferProgressSnapshot` is stateless). Basically, the following would not work if we return a progress with 0% `TrasferProgressSnapshot`
```java
TransferProgress progress = download.progress();
while (!download.completionFuture().isDone()) {
      progress.snapshot().ratioTransferred().ifPresent(System.out::println);
      Thread.sleep(1000);
  }
```
I agree this is not optimal, can you think of other ways to address this?

[the remark this answers is not in the document]
Comment (zoewangg) [reply services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/ObjectTransfer.java]: Discussed offline, turns out I misunderstood it; users are expected to poll `TransferProgress`. Updated to return 0% progress snapshot until it's available.

**Review by dagnir — APPROVED** [review]
